apiserver/admin.go

```diff
@@ -118,83 +105,82 @@ func (a *admin) login(req params.LoginRequest, loginVersion int) (params.LoginRe
+	var facadeFilters []facadeFilterFunc
+	var modelTag string
 	if authResult.anonymousLogin {
-		filters = append(filters, IsAnonymousFacade)
+		facadeFilters = append(facadeFilters, IsAnonymousFacade)
 	}
 	if authResult.controllerOnlyLogin {
-		loginResult.Facades = filterFacades(a.srv.facades, append(filters, IsControllerFacade)...)
-		apiRoot = restrictRoot(apiRoot, controllerFacadesOnly)
+		facadeFilters = append(facadeFilters, IsControllerFacade)
 	} else {
-		loginResult.ModelTag = model.Tag().String()
-		loginResult.Facades = filterFacades(a.srv.facades, append(filters, IsModelFacade)...)
-		apiRoot = restrictRoot(apiRoot, modelFacadesOnly)
+		facadeFilters = append(facadeFilters, IsModelFacade)
+		modelTag = a.root.model.Tag().String()
 	}
 
 	a.root.rpcConn.ServeRoot(apiRoot, serverError)
-
-	return loginResult, nil
+	return params.LoginResult{
+		Servers:       params.FromNetworkHostsPorts(hostPorts),
+		ControllerTag: a.root.model.ControllerTag().String(),
+		UserInfo:      authResult.userInfo,
+		ServerVersion: jujuversion.Current.String(),
+		PublicDNSName: a.srv.publicDNSName(),
+		ModelTag:      modelTag,
+		Facades:       filterFacades(a.srv.facades, facadeFilters...),
+	}, nil
 }
 
 type authResult struct {
-	anonymousLogin      bool
-	userLogin           bool
-	controllerOnlyLogin bool
-	userInfo            *params.AuthUserInfo
+	tag                    names.Tag // nil if external user login
+	anonymousLogin         bool
+	userLogin              bool // false if anonymous user
+	controllerOnlyLogin    bool
+	controllerMachineLogin bool
+	userInfo               *params.AuthUserInfo
 }
 
 func (a *admin) authenticate(req params.LoginRequest) (*authResult, error) {
 	result := &authResult{
 		controllerOnlyLogin: a.root.modelUUID == "",
 		userLogin:           true,
 	}
+	if req.AuthTag != "" {
+		tag, err := names.ParseTag(req.AuthTag)
+		if err != nil {
+			return nil, errors.Annotate(err, "could not parse auth tag")
+		}
+		result.tag = tag
+	}
 
 	// Maybe rate limit non-user auth attempts.
-	machineAgent := false
 	if req.AuthTag != "" {
-		var err error
-		kind, err := names.TagKind(req.AuthTag)
-		// Check for anonymous user login.
-		if kind == names.UserTagKind {
-			userTag, err := names.ParseUserTag(req.AuthTag)
-			if err != nil {
-				return nil, errors.Trace(err)
-			}
-			result.anonymousLogin = userTag.Id() == api.AnonymousUsername && len(req.Macaroons) == 0
-			result.userLogin = !result.anonymousLogin
+		tag, err := names.ParseTag(req.AuthTag)
+		if err == nil {
+			result.tag = tag
 		}
-		if err != nil || kind != names.UserTagKind {
-			addCount := func(delta int64) {
-				atomic.AddInt64(&a.srv.loginAttempts, delta)
-			}
-			addCount(1)
-			defer addCount(-1)
+		if err != nil || tag.Kind() != names.UserTagKind {
+			// Either the tag is invalid, or
+			// it's not a user; rate limit it.
+			atomic.AddInt64(&a.srv.loginAttempts, 1)
+			defer atomic.AddInt64(&a.srv.loginAttempts, -1)
```

Comment:
How does this avoid rate limiting Controller logins. Anything that isn't a User is being limited here.

I suppose the idea is that we still rate limit them, but we just let the Login succeed?

Reply:
Yep, this is just to keep the rate of logins steady. The controllers will still be able to login, they'll just be slowed down. Nothing's changed in this regard (maybe it should later, but I'd rather consider that separately).